pkg/apis/resource/types.go

```diff
@@ -0,0 +1,466 @@
+	// via the name of a Class that was created as part of the
+	// driver deployment.
+	//
+	// The apiserver does not check that the referenced class exists.
+	ClassName string
+
+	// ParametersRef references a separate object with arbitrary parameters
+	// that will be used by the driver when allocating a resource for the
+	// claim.
+	//
+	// The object must be in the same namespace as the Claim.
+	ParametersRef *ClaimParametersReference
+
+	// Allocation can start immediately or when a Pod wants to use the
+	// resource. "WaitForFirstConsumer" is the default.
+	AllocationMode AllocationMode
+}
+
+// AllocationMode describes whether a Claim gets allocated immediately
+// when it gets created (AllocationModeImmediate) or whether allocation is
+// delayed until it is needed for a Pod
+// (AllocationModeWaitForFirstConsumer). Other modes might get added in the
+// future.
+type AllocationMode string
+
+const (
+	// When a Claim has AllocationModeWaitForFirstConsumer, allocation is
+	// delayed until a Pod gets scheduled that needs the Claim. The
+	// scheduler will consider all resource requirements of that Pod and
+	// trigger allocation for a node that fits the Pod.
+	AllocationModeWaitForFirstConsumer AllocationMode = "WaitForFirstConsumer"
+
+	// When a Claim has AllocationModeImmediate, allocation starts
+	// as soon as the Claim gets created. This is done without
+	// considering the needs of Pods that will use the Claim
+	// because those Pods are not known yet.
+	AllocationModeImmediate AllocationMode = "Immediate"
+)
+
+// ClaimStatus tracks whether the resource has been allocated and what
+// the resulting attributes are.
+type ClaimStatus struct {
+	// DriverName is a copy of the driver name from the Class at
+	// the time when allocation started. It's necessary to enable usage of
+	// the claim by kubelet in case that the Class got removed in
+	// the meantime. It also helps the resource driver to determine whether
+	// it needs to handle a claim that got marked for deletion.
+	DriverName string
+
+	// Allocation is set by the resource driver once a resource has been
+	// allocated successfully. Nil indicates that the resource is not
+	// allocated.
+	Allocation *AllocationResult
+
+	// ReservedFor indicates which entities are currently allowed to use
+	// the claim. Usually those are Pods, but other objects are also
+	// possible as long as they exist as resources in the apiserver. The
+	// Claim controller in the kube-controller-manager checks for
+	// stale entries that reference non-existent objects and removes them.
+	//
+	// A scheduler must add a Pod that it is scheduling. This must be done
+	// in an atomic Claim update because there might be multiple
+	// schedulers working on different Pods that compete for access to the
+	// same Claim, the Claim might have been marked for
+	// deletion, or even been deallocated already.
+	//
+	// kubelet will check this before allowing a Pod to run because a a
```

Comment:
Can you make this all clearer.  Pretend to be a user rather than a wizard :)

e.g.

"""
A Pod which references a ResourceClaim which is not reserved for that Pod will not be started.

Reply:
I used exactly that sentence because it seems relevant for end-users.